iceoryx2-ffi/ffi/src/api/publisher.rs

```diff
@@ -10,74 +10,369 @@
+///
+/// # Safety
+///
+/// * `publisher_handle` is valid, non-null and was obtained via [`iox2_cast_publisher_ref_h`]
+/// * `data_ptr` non-null pointer to a valid position in memory
+/// * `data_len` the size of the payload memory
+/// * `number_of_recipients` can be null, otherwise a valid pointer to an [`usize`]
+#[no_mangle]
+pub unsafe extern "C" fn iox2_publisher_send_copy(
+    publisher_handle: iox2_publisher_ref_h,
+    data_ptr: *const c_void,
+    data_len: usize,
+    number_of_recipients: *mut usize,
+) -> c_int {
+    debug_assert!(!publisher_handle.is_null());
+    debug_assert!(!data_ptr.is_null());
+    debug_assert!(data_len != 0);
+
+    let publisher = &mut *publisher_handle.as_type();
+
+    match publisher.service_type {
+        iox2_service_type_e::IPC => send_copy(
+            &publisher.value.as_mut().ipc,
+            data_ptr,
+            data_len,
+            number_of_recipients,
+        ),
+        iox2_service_type_e::LOCAL => send_copy(
+            &publisher.value.as_mut().local,
+            data_ptr,
+            data_len,
+            number_of_recipients,
+        ),
+    }
+}
+
+/// Loans memory from the publishers data segment.
+///
+/// # Arguments
+///
+/// * `handle` obtained by [`iox2_port_factory_publisher_builder_create`](crate::iox2_port_factory_publisher_builder_create)
+/// * `number_of_elements` defines the number of elements that shall be loaned. The elements were
+///    defined via [`iox2_service_builder_pub_sub_set_payload_type_details()`](crate::iox2_service_builder_pub_sub_set_payload_type_details).
+/// * `sample_struct_ptr` - Must be either a NULL pointer or a pointer to a valid [`iox2_sample_mut_t`].
+///    If it is a NULL pointer, the storage will be allocated on the heap.
+/// * `sample_handle_ptr` - An uninitialized or dangling [`iox2_sample_mut_h`] handle which will be initialized by this function call if a sample is obtained, otherwise it will be set to NULL.
+///
+/// Return [`IOX2_OK`] on success, otherwise [`iox2_publisher_loan_error_e`].
+///
+/// # Safety
+///
+/// * `publisher_handle` is valid, non-null and was obtained via [`iox2_cast_publisher_ref_h`]
+/// * The `sample_handle_ptr` is pointing to a valid [`iox2_sample_mut_h`].
+#[no_mangle]
+pub unsafe extern "C" fn iox2_publisher_loan(
+    publisher_handle: iox2_publisher_ref_h,
+    number_of_elements: c_size_t,
+    sample_struct_ptr: *mut iox2_sample_mut_t,
+    sample_handle_ptr: *mut iox2_sample_mut_h,
+) -> c_int {
+    debug_assert!(!publisher_handle.is_null());
+    debug_assert!(!sample_handle_ptr.is_null());
+
+    *sample_handle_ptr = std::ptr::null_mut();
+
+    let mut sample_struct_ptr = sample_struct_ptr;
+    fn no_op(_: *mut iox2_sample_mut_t) {}
+    let mut deleter: fn(*mut iox2_sample_mut_t) = no_op;
+    if sample_struct_ptr.is_null() {
+        sample_struct_ptr = iox2_sample_mut_t::alloc();
+        deleter = iox2_sample_mut_t::dealloc;
+    }
+    debug_assert!(!sample_struct_ptr.is_null());
```

Comment:
Can you please move this into a lambda and call it only when the loan was successful. Else there will be a memory leak.

You can have a look at the subscriber for inspiration.